### PR TITLE
Added tests and modified API

### LINKED ISSUE #19: Fix Github action for unit tests
The .github/workflows/test.yml file needs to be edited so that the tests pass. I removed doctests for now. Having trouble getting it to work so would appreciate any help.

### PR DESCRIPTION
**Pull request by nwlandry**
* Added tests for the following modules
    * connected.py: In the process, I found that the methods were not working correctly so I excluded the s-parameter and eliminated dependency on NetworkX by modifying the BFS algorithm that NetworkX wrote. I excluded `networkx` from the setup.py file because it is no longer needed.
    * hypergraph.py: I added simple tests for the constructor, and properties of the hypergraph. Subhypergraphs and add/delete of edges/nodes is not covered right now.
    * function.py: I added all tests except for attribute get/set methods.
    * linalg.py: This module should be mostly covered.
    * classic.py: Added a simple test here.

Fixes #19, #6, and #9.

The coverage of these tests (roughly speaking) is the following with the `coverage` package:
image

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import os
+import sys
+from pathlib import Path
+import numpy as np
+import pandas as pd
+
+import pytest
+
+
+@pytest.fixture
```
Comment (leotrs) [inline tests/conftest.py:10]: Never seen this done but it's def a good call - I like it.

File: tests/generators/test_classic.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import pytest
```
Comment (leotrs) [inline tests/generators/test_classic.py:1]: Question: do we need to import pytest on each test file? I think not... if that's the case we should probably remove it.

File: tests/generators/test_classic.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+import xgi
+from xgi.exception import XGIError
```
Comment (leotrs) [inline tests/generators/test_classic.py:3]: Nitpick: this import also seems unnecessary.

File: tests/algorithms/test_connected.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
+import pytest
+from tests.conftest import edgelist1
+import xgi
+from xgi.exception import XGIError
+from xgi.readwrite import edgelist
+
+
+def test_is_connected(edgelist1, edgelist3):
+    # edgelist1 = [[1,2,3],[4],[5,6],[6,7,8]]
+    # edgelist2 = [[1,2,3],[3,4],[4,5,6]]
```
Comment (leotrs) [inline tests/algorithms/test_connected.py:10]:
```suggestion
```

File: tests/algorithms/test_connected.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    el1 = edgelist1
+    el2 = edgelist3
+    H1 = xgi.Hypergraph(el1)
+    H2 = xgi.Hypergraph(el2)
```
Comment (leotrs) [inline tests/algorithms/test_connected.py:14]: I believe the first two lines are not achieving anything, no?
```suggestion
    H1 = xgi.Hypergraph(edgelist1)
    H2 = xgi.Hypergraph(edgelist2)
```

File: tests/algorithms/test_connected.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    assert xgi.is_connected(H2) == True
+
+
+def test_connected_components(edgelist1, edgelist2, edgelist4):
+    el1 = edgelist1
+    el2 = edgelist2
+    el4 = edgelist4
+    H1 = xgi.Hypergraph(el1)
+    H2 = xgi.Hypergraph(el2)
+    H3 = xgi.Hypergraph(el4)
```
Comment (leotrs) [inline tests/algorithms/test_connected.py:25]:
```suggestion
    H1 = xgi.Hypergraph(edgelist1)
    H2 = xgi.Hypergraph(edgelist2)
    H3 = xgi.Hypergraph(edgelist4)
```

File: tests/algorithms/test_connected.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    assert sorted([len(cc) for cc in xgi.connected_components(H1)]) == [1, 3, 4]
+    assert {1, 2, 3} in xgi.connected_components(H1)
+    assert {4} in xgi.connected_components(H1)
+    assert {5, 6, 7, 8} in xgi.connected_components(H1)
+    assert sorted([len(cc) for cc in xgi.connected_components(H2)]) == [2, 4]
+    assert {1, 2} in xgi.connected_components(H2)
+    assert {3, 4, 5, 6} in xgi.connected_components(H2)
+    assert sorted([len(cc) for cc in xgi.connected_components(H3)]) == [5]
+    assert {1, 2, 3, 4, 5} in xgi.connected_components(H3)
```
Comment (leotrs) [inline tests/algorithms/test_connected.py:34]: Suggestion: this is both faster and shorter.
```suggestion
    cc1 = xgi.connected_components(H1)
    cc2 = xgi.connected_components(H2)
    cc3 = xgi.connected_components(H3)
    assert sorted([len(cc) for cc in cc1]) == [1, 3, 4]
    assert {1, 2, 3} in cc1
    assert {4} in cc1
    assert {5, 6, 7, 8} in cc1
    assert sorted([len(cc) for cc in cc2]) == [2, 4]
    assert {1, 2} in cc2
    assert {3, 4, 5, 6} in cc2
    assert sorted([len(cc) for cc in cc3]) == [5]
    assert {1, 2, 3, 4, 5} in cc3
```

File: tests/classes/test_function.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+
+
+def test_degree_histogram(edgelist1, edgelist2, edgelist3):
+    H1 = xgi.Hypergraph(edgelist1)
+    H2 = xgi.Hypergraph(edgelist2)
+    H3 = xgi.Hypergraph(edgelist3)
+
+    assert xgi.degree_histogram(H1) == [0, 7, 1]
+    assert xgi.degree_histogram(H2) == [0, 5, 1]
+    assert xgi.degree_histogram(H3) == [0, 4, 2]
```
Comment (leotrs) [inline tests/classes/test_function.py:17]: Suggestion (non-blocking): perhaps in the future we could provide two different functions:
1. `degree_counts` returns what `degree_histogram` currently returns, and
2. `degree_histogram` returns the same data structures as [np.histogram](https://numpy.org/doc/stable/reference/generated/numpy.histogram.html) (i.e. one array with the heights and one array with the bin edges.)

File: tests/classes/test_hypergraph.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    H.name = "test"
+    assert H.name == "test"
+    H = xgi.Hypergraph(name="test")
+    assert H.name == "test"
+
+
+def test_contains(edgelist1):
+    el1 = edgelist1
+    H = xgi.Hypergraph(el1)
+    unique_nodes = list({node for edge in el1 for node in edge})
```
Comment (leotrs) [inline tests/classes/test_hypergraph.py:52]: Nitpick:
```suggestion
    unique_nodes = {node for edge in el1 for node in edge}
```

**Review by leotrs — CHANGES_REQUESTED** [review]
Good stuff all around! I tried to summarize the changes I up to now so that we can write a clean commit  message at the end:
+ Removed networkx dependency
+ Renamed HypergraphError to XGIError
+ General code cleanup
+ Added a bunch of tests yay!

File: tests/algorithms/test_connected.py (hunk 6 above)
Comment (nwlandry) [reply tests/algorithms/test_connected.py:25]: Fixed!

File: tests/algorithms/test_connected.py (hunk 7 above)
Comment (nwlandry) [reply tests/algorithms/test_connected.py:34]: Fixed!

File: tests/algorithms/test_connected.py (hunk 4 above)
Comment (nwlandry) [reply tests/algorithms/test_connected.py:10]: Fixed!

File: tests/algorithms/test_connected.py (hunk 5 above)
Comment (nwlandry) [reply tests/algorithms/test_connected.py:14]: Fixed!

File: tests/classes/test_hypergraph.py (hunk 9 above)
Comment (nwlandry) [reply tests/classes/test_hypergraph.py:52]: Good point. I have no idea why I did it that way.

File: tests/generators/test_classic.py (hunk 3 above)
Comment (nwlandry) [reply tests/generators/test_classic.py:3]: yep! fixed it.

File: tests/generators/test_classic.py (hunk 2 above)
Comment (nwlandry) [reply tests/generators/test_classic.py:1]: Good point

File: tests/classes/test_function.py (hunk 8 above)
Comment (nwlandry) [reply tests/classes/test_function.py:13]: Added as an issue!

**Review by leotrs — APPROVED** [review]

**Pull request merged**